security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/upload_sftp.php

```diff
@@ -0,0 +1,1535 @@
+        $this->processAvailableInput(30, 2);
+        return $this;
+    }
+
+    public function put($local_file, $remote_file = "", $preserve = true)
+    {
+        if (is_file($local_file)) {
+            $this->processAvailableInput();
+            $this->process->put("put " . ($preserve ? "-p " : "")
+                . escapeshellarg($local_file)
+                . (empty($remote_file) ? "" : " " . escapeshellarg($remote_file)));
+            $this->processAvailableInput(30, 2);
+        } else {
+            logger()->info("put: File $local_file doesn't exist.");
+            $this->failed_status = ["file_not_found" => true, "error" => $local_file];
+        }
+
+        return $this;
+    }
+
+    public function mkdir($remote_path)
+    {
+        if (($remote_path = $this->resolve($remote_path)) !== false) {
+            $this->process->put("mkdir " . escapeshellarg($remote_path));
+            $this->processAvailableInput(30, 1);
+        }
+        return $this;
+    }
+
+    public function cd($remote_path)
+    {
+        if (($remote_path = $this->resolve($remote_path)) !== false) {
+            $this->clearError();
+            $this->process->put("cd " . escapeshellarg($remote_path));
+
+            $this->processAvailableInput(30, 1);
+            $error = $this->lastError();
+            $pwd = false;
+            $this->pwd = null;
+
+            if ($error || $remote_path !== ($pwd = $this->pwd())) {
+                $this->failed_status = array_merge(($error ?: []), [
+                    "failure" => true,
+                    "error" => "Failed changing path to '$remote_path' (pwd: '$pwd'); Cause: {$error["error"]}"
+                ]);
+            }
+        }
+        return $this;
+    }
+
+    public function chmod($remote_file, $mode)
+    {
+        if (($remote_file = $this->resolve($remote_file)) !== false) {
+            $this->processAvailableInput();
+            $this->process->put("chmod " . escapeshellarg($mode) . " " . escapeshellarg($remote_file));
+            $this->processAvailableInput(30, 2);
+        }
+        return $this;
+    }
+
+    public function chgrp($remote_file, $group_id)
+    {
+        if (($remote_file = $this->resolve($remote_file)) !== false) {
+            $this->processAvailableInput();
+            $this->process->put("chgrp " . escapeshellarg($group_id) . " " . escapeshellarg($remote_file));
+            $this->processAvailableInput(30, 2);
+        }
+        return $this;
+    }
+
+    public function rm($remote_file)
+    {
+        if (($remote_file = $this->resolve($remote_file)) !== false) {
+            $this->processAvailableInput();
+            $this->process->put("rm " . escapeshellarg($remote_file));
+            $this->processAvailableInput(30, 2);
+        }
+        return $this;
+    }
+}
+
+/**
+ * Utility class for managing SSH host (known_hosts) and identity keys to be used in {@see SftpClient}.
+ */
+class SSHKeys
```

Comment:
please split it into multiple class files in the lib directory.

Example: https://github.com/opnsense/core/pull/3490/files

Reply:
Ok, didn't do it in the first place to have it self contained, but was thinking that the implementations may be useful for something else. 
Do you want me to put it in "mvc/app/library/OPNsense/AcmeClient/..." or "scripts/OPNsense/AcmeClient/library/..."?

Reply:
the mvc one. It will get included automatically so you don't need to require them manually. To run the autoloader, it is usually done by loading 'config.inc' (`require_once("config.inc");`).

Reply:
Ok, thanks a lot for the hint, will do so.

Reply:
Refactoring is done. All generic logic (including file upload) is in class files, all certificate upload related logic is in the script.